introduction_to_amazon_algorithms/jumpstart-foundation-models/text-generation-benchmarking/benchmarking/load_test.py

```diff
@@ -0,0 +1,344 @@
+        self,
+        predictor: Predictor,
+        payload: Dict[str, Any],
+        model_id: str,
+        payload_name: str,
+        tokenizer_model_id: Optional[str] = None,
+        price_per_endpoint: Optional[float] = None,
+    ) -> None:
+        self.predictor = predictor
+        self.payload = payload
+        self.model_id = model_id
+        self.payload_name = payload_name
+        self._logging_prefix = logging_prefix(self.model_id, self.payload_name)
+        if tokenizer_model_id is not None:
+            self.tokenizer = AutoTokenizer.from_pretrained(tokenizer_model_id)
+        else:
+            self.tokenizer = None
+        self.price_per_endpoint = price_per_endpoint
+
+    def predict_once_and_collect_client_results(self) -> PredictionResult:
+        """Perform a single endpoint prediction and produce a PredictionResult."""
+        time_utc_start = datetime.datetime.utcnow()
+        try:
+            result = self.predictor.predict(self.payload, custom_attributes="accept_eula=True")
+        except Exception as e:
+            result = e
+        time_utc_end = datetime.datetime.utcnow()
+        return PredictionResult(time_utc_start, time_utc_end, self.payload, result)
+
+    def run_load_test(self, num_invocations: int, max_workers: int) -> Optional[BatchInvocationStatistics]:
+        """Concurrently invoke an endpoint prediction multiple times and gather results in BatchInvocationStatistics."""
+        time_utc_start = datetime.datetime.utcnow()
+        results = []
+        error = None
+        timeout_seconds = SM_INVOCATION_TIMEOUT * num_invocations / max_workers
+        with futures.ThreadPoolExecutor(max_workers=max_workers) as executor:
+            futures_list = [
+                executor.submit(self.predict_once_and_collect_client_results) for _ in range(num_invocations)
+            ]
+            for future in futures.as_completed(futures_list, timeout=timeout_seconds):
+                result = future.result()
+                if isinstance(result.result, Exception):
+                    error = copy.deepcopy(result.result)
+                    for future_to_cancel in futures_list:
+                        future_to_cancel.cancel()
+                else:
+                    results.append(result)
+
+        if error is not None:
+            raise error
+        time_utc_end = datetime.datetime.utcnow()
+        return BatchInvocationStatistics(time_utc_start, time_utc_end, num_invocations, results)
+
+    def run_latency_load_test(
+        self,
+        num_invocations: int,
+        retry_wait_time: float = RETRY_WAIT_TIME_SECONDS,
+        max_total_retry_time: float = MAX_TOTAL_RETRY_TIME_SECONDS,
+    ) -> Dict[str, Any]:
+        logging.info(f"{self._logging_prefix} Begin latency load test ...")
+        time.sleep(CLOUDWATCH_PERIOD)  # wait for 1 cloudwatch period to ensure no extra queries are reported
+        statistics_latency = self.run_load_test(num_invocations, 1)
```

Comment:
question: shouldn't you wait again here to ensure CW had time to pick up the results?

Reply:
This occurs within the `_extract_cloudwatch_metrics` method. It iteratively retries until all expected statistics are available in CW.